### PR TITLE
kip-21 finality anchor

### PR DESCRIPTION
**Pull request by biryukovmaxim**
Extends the sequencing commitment (`seq_commit`, alias `accepted_id_merkle_root`) so it binds to chain history beyond the active-lanes window. Each block now carries an `inactivity_shortcut`: the `seq_commit` of the highest chain block at `blue_score <= current_blue_score - finality_depth - 1`, i.e. the last chain block just past the live SMT window. Threading this value into `mergeset_context_hash` ties the current seq_commit to a concrete older block whose own seq_commit transitively pins everything before it, so a verifier with only the active window plus the shortcut can still reason about history that has fallen out of the SMT.

The shortcut block hash is resolved by an SMT coinbase-lane lookup first (every chain block touches the coinbase lane, so the latest entry at or below the target depth pins the chain block directly). When the SMT does not retain a deep-enough entry, e.g. immediately after IBD when only the pruning-point window is local, the lookup falls through to a forward walk from a known seed (`selected_parent.inactivity_shortcut_block` when present, otherwise the pruning point) up to the target depth. Both producer and verifier run the same routine, so syncer and syncee agree on the shortcut value regardless of how much SMT history each side has retained.

To make the active-set commitment auditable post-pruning the per-block `SmtBlockMetadata` also stores `payload_root` and `inactivity_shortcut_block`. During pruning-point SMT import the syncee recomputes `seq_commit` from the imported lanes root, the streamed payload root, and the locally derived shortcut, then asserts it against the pruning point header's `accepted_id_merkle_root`.

Splits the anchor compute logic out of the finality module so it depends only on the SMT and block-depth services (and is now driven entirely off `finality_depth`. No fork gate yet, so the shortcut participates in the hash from genesis on every network that runs this branch.

### CODE REVIEW

File: consensus/src/pipeline/virtual_processor/processor.rs [hunk 1]
```diff
@@ -747,31 +758,104 @@ impl VirtualStateProcessor {
...
+            return ZERO_HASH;
+        }
+
+        let search_from = if let Some(sp_inactivity_shortcut_block) =
+            self.smt_metadata_store.get(selected_parent).map(|md| md.inactivity_shortcut_block).optional().unwrap()
+            && sp_inactivity_shortcut_block != ZERO_HASH
+        {
+            sp_inactivity_shortcut_block
+        } else {
+            self.headers_store.get_header(selected_parent).unwrap().pruning_point
```
Comment (michaelsutton) [inline consensus/src/pipeline/virtual_processor/processor.rs]: I'm not sure this is a correct fallback. Why can't we assume that sp has a valid `inactivity_shortcut_block` ?

File: consensus/src/pipeline/virtual_processor/processor.rs [hunk 2]
```diff
@@ -747,31 +758,104 @@ impl VirtualStateProcessor {
...
+            current = chain_block;
+        }
+        current
+    }
+
+    /// Derive the `inactivity_shortcut` value committed in
+    /// `MergesetContext.inactivity_shortcut` from its block hash. `ZERO_HASH`
+    /// (no block known) and genesis both yield `ZERO_HASH`.
+    pub(super) fn inactivity_shortcut(&self, inactivity_shortcut_block: Hash) -> Hash {
+        if inactivity_shortcut_block == ZERO_HASH || inactivity_shortcut_block == self.genesis.hash {
```
Comment (michaelsutton) [inline consensus/src/pipeline/virtual_processor/processor.rs]: `header(genesis).accepted_id_merkle_root` is always zero hash (see https://github.com/michaelsutton/rusty-kaspa/blob/4f37a4a705c83992dbc8bbbd2a4c470ee0ef2c2d/consensus/core/src/config/genesis.rs#L36)

**Comment by demisrael** [comment]
## Review of PR #17 — kip-21 finality anchor

Thanks for the well-shaped extension @biryukovmaxim. Source-walked at HEAD `c3c731aa` against base `michaelsutton:zk-updates@3201436e`. The PR is structurally sound; one consensus-crash hazard merits attention before merge.

### What checks out

- **`mergeset_context_hash` threading is correctly length-distinguishing.** The 24-byte preimage (pre-activation, `None`) and the 56-byte preimage (post-activation, `Some`) cannot collide under BLAKE3's chunk-tree finalization. The KIP-21 §6.5.1 three-field preimage is preserved bit-for-bit when the flag is off — pre-zk-hardening peers continue producing KIP-21-conformant context hashes.

- **Transitive pinning holds** under the "every chain block has a coinbase tx with `subnetwork_id == SUBNETWORK_ID_COINBASE`" invariant. Coinbase-lane SMT lookup returns the latest canonical entry at or below `target_bs`; `is_smt_canonical` at `processor.rs:785` correctly filters reorg-stale entries. Worth noting: pinning extends back only to the toccata activation boundary — pre-toccata blocks set `accepted_id_merkle_root` via the legacy `calc_accepted_id_merkle_root` path, which doesn't chain through the seq_commit recurrence. PR body could state that bound explicitly.

- **Producer-verifier gate symmetry is clean.** `compute_seq_commit`, `recompute_seq_commit`, and the IBD mirror in `protocol/flows/src/ibd/flow.rs::sync_new_smt_state` evaluate `zk_hardening_activation.is_active(<block's daa_score>)` with identical fold semantics — three independent paths converge on the same predicate. The genesis-context-hash fix in `a7cfe844` was necessary; without it, simnet / simpa would have diverged at block 1.

- **`lanes_root` / `payload_root` / `parent_seq_commit` / `seq_commit` composition** is cryptographically sound under standard BLAKE3 keyed-mode security (≈2^128 collision-resistance, ≈2^256 preimage). All ten SeqCommit-family domain tags in `crypto/hashes/src/hashers.rs::blake3_hasher!` derive distinct keys → independent PRFs. Forging accepted-by-verifier IBD wire payloads reduces to a BLAKE3 keyed-mode collision.

### High-priority before merge

#### 1. Panic-on-`ToccataV0` is reachable at the activation boundary

`consensus/src/model/stores/smt_metadata.rs:41` and `:48` both `panic!(...)` on the `ToccataV0` arm (legacy 40-byte layout lacks `payload_root` and `inactivity_shortcut_block`). The doc at lines 24-25 claims "the upgrade boundary is enforced upstream, so callers reach for a field only after migration is complete" — I couldn't locate that enforcement (no IBD-trigger reformatting, no startup migration pass, no in-tree V0→V1 backfill).

Two callers reach the accessor without guarding the variant:

- `compute_inactivity_shortcut_block` at `processor.rs:825-830` — the Path-2 fallback's `.map(|md| md.inactivity_shortcut_block())` runs on V0 rows before `unwrap_or` has a chance.
- `get_parent_smt_metadata` at `processor.rs:864-873` — same shape, `.map(|meta| (meta.active_lanes_count(), meta.inactivity_shortcut_block()))` panics inside the closure.

Reachable scenario: a node that ran pre-KIP-21 code against the Toccata branch has V0 rows accumulated; after upgrading to the post-KIP-21 binary, the first toccata-active block whose SP has a V0 row triggers the panic. On `testnet-12` (toccata-always-active), this is the first block post-upgrade for any node with pre-existing chain state.

Adjacent failure mode worth fixing at the same time: when `smt_metadata_store.get(selected_parent)` returns `Err` (no row at all — possible at the pre-toccata → post-toccata + post-zk-hardening boundary), `unwrap_or(selected_parent)` seeds the forward walk at `selected_parent`, the loop iterates zero blocks, and the function returns `selected_parent` whose `bs > target_bs` — violating the documented invariant. Same root: incomplete migration story. Returning `genesis.hash` from this branch is the simplest safe fix (folds to `ZERO_HASH` downstream).

Remediation options (your call):

1. Change `payload_root()` / `inactivity_shortcut_block()` to return `Option<Hash>` and have callers handle absence explicitly.
2. Treat `ToccataV0` as "no recorded shortcut block" — fold to `ZERO_HASH` matching the `KeyNotFound` branch.
3. Eager startup-time migration pass that rewrites V0 rows to V1.
4. Document and runtime-check an activation-ordering invariant.

A test fixture can be built using the existing `legacy_row_key` helper at `smt_metadata.rs:130-134`; constructing a hand-crafted 40-byte row for the SP of a block whose `compute_seq_commit` enters the Path-2 fallback would have caught this pre-merge. The current activation-crossing tests in `simpa/tests/smt_repro.rs` and `virtual_processor/tests.rs` always start with V1 fixtures.

#### 2. Production-path `assert!` on lane-count invariant

`protocol/flows/src/v11/request_pruning_point_smt_state.rs:139`:

```rust
assert!(lanes_sent == reader_count && lanes_sent == expected_count);
```

Always-on (release + debug) on a production code path. Realistic triggers don't require wire-level peer attack:

1. Local-state inconsistency: on-disk SMT lane count diverges from stored `active_lanes_count` (pruning races, metadata write completing after count changed).
2. Race against pruning processor: `open_pruning_point_smt_lane_stream` after `metadata.active_lanes_count` capture; no locks in the changed region prevent this.
3. Reader early-exit via `tx.blocking_send(…).is_err()` at lines 90-92 returns `Ok(count)` with a partial value; the async loop completes with `lanes_sent == reader_count` (both saw partial set) but `expected_count` is the original captured value.

Same shape as `ProtocolError::Other(…)` at lines 51-54 / 79-81 of the same file would be appropriate — surface as recoverable error rather than crashing the node.

### Cross-anchored Toccata-track concerns

**`SmtBlockMetadata::ToccataV0`/`ToccataV1` variant naming ties to the wrong fork.** The on-disk V0→V1 boundary is the moment this PR ships (when `payload_root` + `inactivity_shortcut_block` start being persisted), not `toccata_activation`. `zk_hardening_activation` set to `never()` still produces V1 rows — the schema change is unconditional; only the gate's `mergeset_context_hash` consumption is conditional. Renaming to `WithoutAnchor` / `WithAnchor` or `LegacyKip21` / `KipFinalityAnchor` would match the actual semantic boundary. Same root as H1.

**P2P version bump 10 → 11 hard-rejects v10 peers.** `protocol/flows/src/flow_context.rs:837-849` matches v11+, v9, v8, v7 explicitly; v10 falls into the rejection arm. `testnet-12` advertises v10 today on Toccata; operators of v10 nodes will lose their peers on the upgrade step. The `TN12_LAUNCH_PROTOCOL_VERSION = 9` promotion shim at lines 819-821 silently treats unupgraded TN12 peers as v11-capable — but their actual wire bytes are 96-byte (V0) `SmtMetadataMessage.data`, which the 128-byte decoder at `ibd/streams.rs:241-262` rejects. Promotion shim becomes a coordination liability after this PR ships; worth removing or gating on a separate compat predicate. TN12 needs a coordinated all-node upgrade.

**Other Toccata interactions check out.** `payload_root` folds in the same `block_txs[0].payload` bytes both producer and verifier see — Toccata's mass-pricing / standardness changes don't touch coinbase-payload encoding. Always-active networks (tn-12, simnet, simpa) are self-consistent at genesis via the clamp + `inactivity_shortcut(genesis) → ZERO_HASH` + `compute_genesis_accepted_id_digests` gating. Toccata-boundary-straddling pruning points aren't a hazard at the PP level since a PP is a single block (pre-toccata PPs skip SMT sync at `ibd/flow.rs:691-694`); chain-segments crossing toccata during IBD are untested in default CI (simpa fixtures are always-active or always-not-active only) — worth a fixture variant if you want belt-and-braces.

### Spec / docs / tests

- **KIP-21 doesn't actually specify `inactivity_shortcut` or `payload_root`.** PR #36 on `kaspanet/kips` covers the underlying seq_commit construction but not this PR's finality-anchor extension. PR body is de-facto spec. Either extend KIP-21 to cover the anchor layer, or ship as KIP-21a / sibling KIP, so downstream implementations have a normative document.
- **PR body claim "No fork gate yet"** is stale post `f2ed6eca` — quick rewrite to reflect current state.
- **`SmtBlockMetadata` doc claim "the upgrade boundary is enforced upstream" is unenforced** at `smt_metadata.rs:24-25` — either implement enforcement (H1 remediation) or rewrite to name the actual contract.
- **`test_mergeset_context_hash_golden`** at `hashing.rs:240-270` declares `let expected = Hash::from_bytes([0xAA; 32]); let _ = expected;` (dead) with a trailing comment promising pinning that wasn't restored. Either pin actual bytes (matching `test_lane_key_golden` / `test_seq_commit_golden`) or rename to `_anchor_sensitivity` and delete scaffolding.
- **`daemon_zk_hardening_activation_test` is `#[ignore]`-gated** at `daemon_integration_tests.rs:143`. The simpa test at `smt_repro.rs:69-71` references it for pre-toccata crossing coverage; gated-off means the chain has a hole.
- **`compute_seq_commit` and `recompute_seq_commit`** share their entire steady-state body. Drift surface for the parity contract `daemon_zk_hardening_activation_test` is supposed to catch. Shared helper parameterized on producer-vs-verifier eliminates the drift.
- **`inactivity_shortcut_for` test helper** at `consensus_integration_tests.rs:1492-1502` duplicates `VirtualStateProcessor::inactivity_shortcut`. Same drift hazard.
- **`_parent_shortcut`** from `get_parent_smt_metadata` is unused at both call sites — drop or wire (consuming it would also resolve the H1 path partly).

### Hygiene

A pre-merge interactive rebase would help: squash `59cf0ea1` (revert) + `2edc2a53` (removal) into foundation `cbf53640`; fold `fbc8db53 fmt` into the commit it formats; rebase or squash merged sub-PRs `af00626d` (#22) + `abfa5c65` (#23) into their semantic landing commits to keep linear history.

### Future hardening (not blocking)

`SeqCommitMerkleBranch` is the keyed domain for five structural roles: Merkle inner branches in `activity_digest_lane`, in `miner_payload_root`, plus the 2-element folds in `payload_and_context_digest`, `seq_state_root`, `seq_commit`. Sound under the current 32-byte input invariant. Brittle under future extension: adding a third field to `seq_commit` (a hypothetical `next_anchor` 32-byte field with naive `H_seq(P || S || N)`) would share a key family with `activity_digest_lane` over a 3-leaf tree. Per-role tags (`SeqCommitFinalize`, `SeqCommitStateRoot`, `SeqCommitPayloadCtxDigest`) on a follow-on hardening cycle would remove the structural coupling.

### Summary

`cargo fmt --all -- --check` and `cargo clippy --workspace --tests --benches --examples -- -D warnings` both pass at HEAD on stable 1.95.0. CI-pinned is 1.93.0; newer clippy is generally stricter, so PASS at 1.95.0 implies PASS at 1.93.0 modulo any lint removed between versions (rare) — worth a re-run on 1.93.0 before merge for belt-and-braces.

Highest-priority items to address before merge:
1. **H1**: close the `ToccataV0` panic + no-row degenerate seed at `processor.rs:824-830` / `smt_metadata.rs:41,48`.
2. **M1**: replace production-path `assert!` at `request_pruning_point_smt_state.rs:139` with `ProtocolError::Other`.
3. Spec + PR-body updates.
4. Test coverage gap on V0 → V1 migration + drop `#[ignore]` on the daemon activation-crossing test.

For michaelsutton's open threads: the forward-walk fallback IS necessary (post-IBD and post-reorg cases need real chain advancement, not just trust of SP's recorded value); the genesis edge case is closed by `c3c731aa` with `inactivity_shortcut_block_clamps_to_genesis_within_finality_depth` pinning it.

Happy to go deeper on any of the above if useful.

File: consensus/smt-store/src/processor.rs [hunk 3]
```diff
@@ -539,9 +547,12 @@ pub struct SmtBuild {
     pub root: Hash,
     node_changes: SmtNodeChanges,
     lane_changes: BlockLaneChanges,
-    /// Set by `build_seq_commit` after computing the seq_commit components.
     pub payload_and_ctx_digest: Hash,
     pub active_lanes_count: u64,
+    /// KIP-21: block hash whose seq_commit IS the `inactivity_shortcut`. Kept here
+    /// for the per-block parent-fallback path in `compute_inactivity_shortcut_block`;
+    /// it does NOT feed `mergeset_context_hash` anymore.
```
Comment (michaelsutton) [inline consensus/smt-store/src/processor.rs]: leftover comment

File: consensus/seq-commit/src/hashing.rs [hunk 4]
```diff
@@ -1,24 +1,13 @@
 //! Hash functions for sequencing commitments.
```
Comment (michaelsutton) [inline consensus/seq-commit/src/hashing.rs:1]: can we add something like this ascii here to describe the hashing structure?:

```
SeqCommit(B)
└── H_seq
    ├── parent_seq_commit = SeqCommit(selected_parent(B))
    └── state_root
        └── H_seq
            ├── activity_root
            │   └── H_activity_root
            │       ├── inactivity_shortcut
            │       └── smt_root
            │           └── SMT root over active lanes
            │
            └── payload_and_ctx_digest
                └── H_seq
                    ├── context_hash
                    └── payload_root
```

File: testing/integration/src/daemon_integration_tests.rs [hunk 5]
```diff
@@ -131,6 +131,63 @@ async fn daemon_toccata_activation_log_file_test() {
     );
 }
 
+/// KIP-21 zk_hardening_activation crossing: pre-activation blocks commit
```
Comment (michaelsutton) [inline testing/integration/src/daemon_integration_tests.rs:134]: in addition to this, can we add a simpa hardening activation at daa 2000? (in `test_pruning_via_simpa`)

File: database/Cargo.toml [hunk 6]
```diff
@@ -16,7 +16,7 @@ faster-hex.workspace = true
 indexmap.workspace = true
 itertools.workspace = true
 kaspa-hashes.workspace = true
-kaspa-utils = { workspace = true, features = ["mem_size"]}
+kaspa-utils = { workspace = true, features = ["mem_size", "fd_budget"]}
```
Comment (michaelsutton) [inline database/Cargo.toml:19]: needed?

File: protocol/p2p/proto/p2p.proto [hunk 7]
```diff
@@ -318,7 +318,10 @@ message RequestPruningPointSmtStateMessage {
 }
 
 message SmtMetadataMessage {
-  bytes data = 1;                  // 96 bytes: lanes_root || payload_and_ctx_digest || parent_seq_commit
+  // Dual schema (variant selected by length):
```
Comment (michaelsutton) [inline protocol/p2p/proto/p2p.proto]: leftover comment change
